src/uct/gaudi/base/gaudi_topo.c

```diff
@@ -0,0 +1,1919 @@
+    if ((device1 == UCS_SYS_DEVICE_ID_UNKNOWN) ||
+        (device2 == UCS_SYS_DEVICE_ID_UNKNOWN) || (device1 == device2)) {
+        *distance = (ucs_global_opts.dist.node.bandwidth > 0) ?
+                            ucs_global_opts.dist.node :
+                            gaudi_fallback_node_distance;
+        return UCS_OK;
+    }
+
+    /* Default distance */
+    *distance = ucs_topo_default_distance;
+
+    /* Step 1: Check NUMA nodes */
+    numa1 = ucs_topo_sys_device_get_numa_node(device1);
+    numa2 = ucs_topo_sys_device_get_numa_node(device2);
+
+    if (numa1 != UCS_NUMA_NODE_UNDEFINED && numa2 != UCS_NUMA_NODE_UNDEFINED) {
+        /* Different NUMA nodes */
+        if (numa1 != numa2) {
+            *distance = (ucs_global_opts.dist.sys.bandwidth > 0) ?
+                                ucs_global_opts.dist.sys :
+                                gaudi_fallback_sys_distance;
+            return UCS_OK;
+        }
+        /* Same NUMA; continue to PCIe refinement */
+    }
+
+    /* Step 2: Same NUMA or NUMA unknown or undefined */
+
+    /* allocate three scratch buffers */
+    status = ucs_string_alloc_path_buffer(&path1, "path1");
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    status = ucs_string_alloc_path_buffer(&path2, "path2");
+    if (status != UCS_OK) {
+        goto free_path1;
+    }
+
+    status = ucs_string_alloc_path_buffer(&common_path, "common_path");
+    if (status != UCS_OK) {
+        goto free_path2;
+    }
+
+    /* convert devices to sysfs paths */
+    status = uct_gaudi_sys_dev_to_sysfs_path(device1, path1, PATH_MAX);
+    if (status != UCS_OK) {
+        goto free_all;
+    }
+
+    status = uct_gaudi_sys_dev_to_sysfs_path(device2, path2, PATH_MAX);
+    if (status != UCS_OK) {
+        goto free_all;
+    }
+
+    hops = uct_gaudi_count_pcie_hops(path1, path2, common_path);
+
+    /* Note: DO NOT treat hops==255 as system distance. In this topology,
+     * Gaudi/NIC can share NUMA but different root complexes. The 255 hop
+     * penalty naturally degrades the score without overriding NUMA affinity,
+     * which is the correct behavior.
+     */
+    distance->latency = (ucs_global_opts.dist.node.bandwidth > 0 ?
+                                 ucs_global_opts.dist.node.latency :
+                                 gaudi_fallback_node_distance.latency) +
+                        hop_latency_ns * hops;
+
+    distance->bandwidth = (ucs_global_opts.dist.node.bandwidth > 0 ?
+                                   ucs_global_opts.dist.node.bandwidth :
+                                   gaudi_fallback_node_distance.bandwidth) *
+                          pow(hop_bw_penalty, hops);
+
+    status = UCS_OK;
+
+free_all:
+    ucs_free(common_path);
+free_path2:
+    ucs_free(path2);
+free_path1:
+    ucs_free(path1);
+out:
+    return status;
+}
+
+/* Create connection matrix */
+static ucs_status_t uct_gaudi_create_connection_matrix()
```

Comment:
general question: this seems like some duplication of the code in topo.c. so what is missing in the current topology logic for gaudi use case?